chaos_genius/alerts/__init__.py

```diff
@@ -43,40 +48,42 @@ def check_and_trigger_alert(alert_id):
     elif (
         alert_info.alert_type == "KPI Alert" and alert_info.kpi_alert_type == "Anomaly"
     ):
-        anomaly_obj = AnomalyAlertController(alert_info.as_dict)
-        return anomaly_obj.check_and_prepare_alert()
+        anomaly_obj = AnomalyAlertController(alert_info)
+        return anomaly_obj.check_and_send_alert()
     elif alert_info.alert_type == "KPI Alert" and alert_info.kpi_alert_type == "Static":
-        static_kpi_alert = StaticKpiAlertController(alert_info.as_dict)
+        StaticKpiAlertController(alert_info.as_dict)
 
     return True
 
 
 def trigger_anomaly_alerts_for_kpi(
-    kpi_obj: Kpi, end_date: date
-) -> Tuple[List[int], List[int]]:
+    kpi_obj: Kpi,
+) -> Tuple[List[int], List[Tuple[int, Exception]]]:
     """Triggers anomaly alerts starting from end_date.
 
     Args:
         kpi_obj (Kpi): Object of kpi for which alerts are to be triggered
-        end_date (dateimte.datetime): Datetime object containing the upper bound of anomaly date values
+        end_date (dateimte.datetime): Datetime object containing the upper bound of
+            anomaly date values
 
     Returns:
         List[int]: List of alert IDs for which alert messages were successfully sent
-        List[int]: List of alert IDs for which alert failed
+        List[Tuple[int, Exception]]: List of alert IDs and exceptions for which alert
+            failed
     """
-    success_alerts = []
-    errors = []
-    alerts = Alert.query.filter(
-        Alert.kpi == kpi_obj.id, Alert.active == True, Alert.alert_status == True
+    success_alerts: List[int] = []
+    errors: List[Tuple[int, Exception]] = []
+    alerts: List[Alert] = Alert.query.filter(
+        Alert.kpi == kpi_obj.id,
+        Alert.active == True,  # noqa: E712
+        Alert.alert_status == True,  # noqa: E712
     ).all()
     for alert in alerts:
         try:
-            anomaly_obj = AnomalyAlertController(
-                alert.as_dict, anomaly_end_date=end_date
-            )
-            anomaly_obj.check_and_prepare_alert()
+            anomaly_obj = AnomalyAlertController(alert)
+            anomaly_obj.check_and_send_alert()
             success_alerts.append(alert.id)
         except Exception as e:
             logger.error(f"Error running alert for Alert ID: {alert.id}", exc_info=e)
-            errors.append(alert.id)
+            errors.append((alert.id, e))
```

Comment:
The issue here is that there can be multiple alerts for a KPI. In the function that calls this, we're checkpointing just the first error (if it exists). If we have a decorator on `check_and_send_alert()` it would cause multiple `Alert trigger` checkpoints for a single KPI. This would prevent some further improvements in task monitor - for example, there can't be a progress bar shown for a KPI since there will be a variable number of checkpoints for a KPI.

It's a good suggestion. We should think about it when improving the task monitor.